[P/D]`kv_output_aggregator` support heterogeneous

LCAIZJ opened the pull request:
## Purpose
1. D is memory bound so greater TP size will yield better performance. However, in the MLA scenario, each TP worker maintains the full kvcache, which prevents actual GPU memory savings. As a result, when MLA is enabled, the TP size for P often exceeds that of D.
2. When the TP size for P exceeds that of D, the current mechanism of kv_output_aggregator fails to function correctly. As illustrated in the figure below, both finished_sending and finished_recving should be 2 in this scenario. However, the kv_output_aggregator incorrectly initializes them to world_size, causing finished_sending to remain at 4 and preventing the release of P's kvcache.
<img width="440" height="462" alt="image" src="https://github.com/user-attachments/assets/9ca0a4ca-1309-4a74-8326-dcf858b0c9db" />

3. To resolve this issue, we introduce a new interface get_finished_count() in the connector's base class. This method is implemented by specific connector subclasses to return the correct values for finished_sending and finished_recving. When different executors invoke get_finished_count() to initialize the kv_output_aggregator, it ensures the counters are configured correctly.

## Test Plan
Model：DeepSeek-R1
Distributed Strategy：1P(TP16)1D(DP4、TP4)

## Test Result
Successfully processed 10,000 benchmark prompts with 4K input tokens and 1.5K output tokens per request. 
<img width="1333" height="546" alt="image" src="https://github.com/user-attachments/assets/cf56a983-0c5f-4730-a2eb-f6b37b3b7a1d" />
Since the primary goal was to verify overall functionality, we did not enable many optimization features during benchmark execution. As a result, the TTFT and TPOT metrics are suboptimal, though the verification of the core functionality was successfully completed.


---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

Commit:
fix kv_output_aggregator support heterogeneous

Co-authored-by: jianzs <zheng.shoujian@outlook.com>
Co-authored-by: baxingpiaochong <771405853@qq.com>
Co-authored-by: CalvinXKY <kyxiezju@163.com>
Co-authored-by: Pz1116 <zpbzpb123123@gmail.com>

Signed-off-by: LCAIZJ <leichao139636@163.com>

Commit:
fix get_finished_count return None

Signed-off-by: LCAIZJ <leichao139636@163.com>

Commit:
executor base add init_kv_output_aggregator

Signed-off-by: LCAIZJ <leichao139636@163.com>

LCAIZJ committed:
fix yapf

Signed-off-by: LCAIZJ <leichao139636@163.com>

LCAIZJ committed:
fix ruff check

Signed-off-by: LCAIZJ <leichao139636@163.com>

NickLucche reviewed (changes requested):
Thanks a lot for contributing to this @LCAIZJ !

One big issue with this interface is that heterogeneous TP is, for some connectors like Nixl but I would argue more broadly for discovery-based ones, a "runtime property", as the connector only finds out about the heterogeneous state after the first handshake (assuming "static" P configuration on the other side).
Hence here we wouldn't be able to return a sensible `get_finished_count` at init time.

>Successfully processed 10,000 benchmark prompts with 4K input tokens and 1.5K output tokens per request.

What connector did you use for this, as I don't see any concrete `get_finished_count` impl on this PR?

Also, it would be nice if we could add some minimal unit tests, perhaps testing the example you reported.  

---
PS 
 > When the TP size for P exceeds that of D, the current mechanism of kv_output_aggregator fails to function correctly

This is also partly the reason why in disagg PD settings with NixlConnector we do not allow P TP size to exceed D's https://github.com/vllm-project/vllm/blob/main/vllm/distributed/kv_transfer/kv_connector/v1/nixl_connector.py#L909.
But I would like to understand, how popular/important is this MLA+P TP>D PT use-case, rather than assuming MoE and going wide EP?
I feel maintaining both P-TP</>D-TP in NixlConnector might complicate code quite a bit.
Nothing against supporting that for other connectors though ofc.

LCAIZJ commented on vllm/v1/engine/core.py:
```diff
@@ -128,6 +128,7 @@ def __init__(self,
             log_stats=self.log_stats,
         )
         self.use_spec_decode = vllm_config.speculative_config is not None
+        self.model_executor.init_kv_output_aggregator()
```
We had initially intended to implement this approach as well, but encountered constraints. Since the connector is initialized within the scheduler, and the scheduler itself must be initialized after the executor due to dependency constraints, we ultimately invoke the executor's `init_kv_output_aggregator` method after scheduler initialization completes.

LCAIZJ commented on vllm/executor/executor_base.py:
```diff
@@ -252,6 +256,17 @@ async def check_health_async(self) -> None:
+            self.kv_output_aggregator = KVOutputAggregator(
+                kv_connector.get_finished_count()
+                or self.parallel_config.world_size)
+        else:
+            self.kv_output_aggregator = KVOutputAggregator(
+                self.parallel_config.world_size)
```
I agree with your suggestion. The current implementation follows the previous pattern where aggregator initialization didn't check for kv transfers (https://github.com/vllm-project/vllm/blob/main/vllm/v1/executor/multiproc_executor.py#L127). I also believe there's no need to initialize the aggregator when no kv transfers are happening - I'll remove this logic.

LCAIZJ committed:
not to initialize the aggregator when no kv transfers are happening

Signed-off-by: LCAIZJ <leichao139636@163.com>

youkaichao commented:
> how popular/important is this MLA+P TP>D PT use-case

@NickLucche  For MLA, people usually do Decode TP=1, and Prefill TP > 1.

NickLucche reviewed (approved):
Hey @LCAIZJ , thanks for adding valuable context to this PR.
To re-iterate my previous comment, although I understand one could have the information about the tp size of other instances readily available into the kv transfer config (as mooncake does), this is just not the way NixlConnector works right now, because general config doesn't flow top to bottom with llm-d.
Therefore, NixlConnector will simply not implement this method for the time being, as it doesn't have a way to tell whether this is a heterogeneous TP case at init time.  

Other than clarifying this point, I am fine with this PR in order to enable Mooncake+other non discovery-based connectors to perform at best.
 
We should just change the title of this PR and remove `[NIXL]`.

---

Re: Prefill TP > Decode TP thanks for answering my curiosity (this is really orthogonal to this PR) . ~To understand the need for that better (nixl-side), @youkaichao are you referring to a pure P/D setup with a general MoE+MLA model, not following the Wide-EP pattern?~ Synched with Kaichao offline 👍🏻

LCAIZJ committed:
delete KVOutputAggregator import in multiproc_executor

Signed-off-by: LCAIZJ <leichao139636@163.com>

LCAIZJ committed:
Trigger CI pipeline

Signed-off-by: LCAIZJ <leichao139636@163.com>

LCAIZJ commented:
> Hey @LCAIZJ , thanks for adding valuable context to this PR. To re-iterate my previous comment, although I understand one could have the information about the tp size of other instances readily available into the kv transfer config (as mooncake does), this is just not the way NixlConnector works right now, because general config doesn't flow top to bottom with llm-d. Therefore, NixlConnector will simply not implement this method for the time being, as it doesn't have a way to tell whether this is a heterogeneous TP case at init time.
> 
> Other than clarifying this point, I am fine with this PR in order to enable Mooncake+other non discovery-based connectors to perform at best.
> 
> We should just change the title of this PR and remove `[NIXL]`.
> 
> Re: Prefill TP > Decode TP thanks for answering my curiosity (this is really orthogonal to this PR) . ~To understand the need for that better (nixl-side), @youkaichao are you referring to a pure P/D setup with a general MoE+MLA model, not following the Wide-EP pattern?~ Synched with Kaichao offline 👍🏻

Hi @NickLucche ,

Thank you for your feedback! If the current modifications address the TP heterogeneity issue in a way that benefits other connectors, I believe the changes are solid.

I’ve rebased the code and removed [nixl] from the title as requested. Could we proceed with merging this now?

LCAIZJ committed:
fix

Signed-off-by: LCAIZJ <leichao139636@163.com>

LCAIZJ committed:
delete no used import

Signed-off-by: LCAIZJ <leichao139636@163.com>

LCAIZJ committed:
Trigger CI pipeline

Signed-off-by: LCAIZJ <leichao139636@163.com>

LCAIZJ committed:
Trigger CI pipeline

Signed-off-by: LCAIZJ <leichao139636@163.com>

LCAIZJ committed:
fix

Signed-off-by: LCAIZJ <leichao139636@163.com>

LCAIZJ committed:
fix yapf

Signed-off-by: LCAIZJ <leichao139636@163.com>

NickLucche reviewed (changes requested):
Sorry, just left another comment after the recent changes which I had missed earlier.

LCAIZJ committed:
avoid modifying the scheduler

Signed-off-by: LCAIZJ <leichao139636@163.com>

LCAIZJ committed:
#suppress-bc-linter

Signed-off-by: LCAIZJ <leichao139636@163.com>

LCAIZJ committed:
Trigger CI pipeline #suppress-bc-linter

Signed-off-by: LCAIZJ <leichao139636@163.com>

LCAIZJ committed:
#suppress-bc-linter

Signed-off-by: LCAIZJ <leichao139636@163.com>

LCAIZJ commented:
@NickLucche All CI pipelines have passed. Would you mind reviewing the PR again when you have time?

NickLucche commented:
This is likely the most minimal enabling set of changes now, thank you!
